metagraph/src/graph/representation/canonical_dbg.cpp

```diff
@@ -577,5 +564,171 @@ void CanonicalDBG::reverse_complement(std::string &seq,
     std::swap(path, rev_path);
 }
 
+
+void CanonicalDBG
+::adjacent_outgoing_from_rc(node_index node,
+                            const std::function<void(node_index)> &callback,
+                            const std::string &spelling_hint) const {
+    //        lshift    rc
+    // AGCCAT -> *AGCCA -> TGGCT*
+    if (const auto *dbg_succ_ = get_dbg_succ(*graph_)) {
+        //   AGAGGATCTCGTATGCCGTCTTCTGCTTGAG
+        //-> AGAGGATCTCGTATGCCGTCTTCTGCTTGA
+        //-> TCAAGCAGAAGACGGCATACGAGATCCTCT
+        const boss::BOSS &boss = dbg_succ_->get_boss();
+        auto *cache = get_extension_threadsafe<NodeFirstCache>();
+        if (!cache)
+            cache = fallback_cache_.get();
+
+        bool is_dummy = (spelling_hint.back() == boss::BOSS::kSentinel);
+
+        boss::BOSS::edge_index rc_edge = cache->get_prefix_rc(
+            dbg_succ_->kmer_to_boss_index(node),
+            spelling_hint
+        );
+
+        if (!rc_edge)
+            return;
+
+        boss.call_outgoing(rc_edge, [&](boss::BOSS::edge_index adjacent_edge) {
+            assert(dbg_succ_);
+            node_index prev = dbg_succ_->boss_to_kmer_index(adjacent_edge);
+            if (prev != DeBruijnGraph::npos
+                    && (!is_dummy || boss.get_W(adjacent_edge) != boss::BOSS::kSentinelCode)) {
+                callback(prev);
+            }
+        });
+
+        return;
+    }
+
+    // Do the checks by directly mapping the sequences of the desired k-mers.
+    // For non-DBGSuccinct graphs, this should be fast enough.
+    std::string rev_seq = graph_->get_node_sequence(node).substr(0, graph_->get_k() - 1);
+    ::reverse_complement(rev_seq.begin(), rev_seq.end());
+    rev_seq.push_back('\0');
+
+    const auto &alphabet = graph_->alphabet();
+    for (size_t c = 0; c < alphabet.size(); ++c) {
+        if (alphabet[c] != boss::BOSS::kSentinel) {
+            rev_seq.back() = complement(alphabet[c]);
+            node_index prev = graph_->kmer_to_node(rev_seq);
+            if (prev != DeBruijnGraph::npos)
+                callback(prev);
+        }
+    }
+}
+
+void CanonicalDBG
+::adjacent_incoming_from_rc(node_index node,
```

Comment:
How this is "incoming" I still don't get... This is confusing

Reply:
Add the comment above `// ATGGCT -> TGGCT* -> *AGCCA`
Without this comment, it's impossible to guess what this function does

Reply:
I added some extra comments

Reply:
Also, I renamed these functions so the header looks like this
```
    // Find incoming nodes that are on the reverse complement strand of node
    void adjacent_incoming_rc_strand(node_index node,
                                     const std::function<void(node_index)> &callback,
                                     const std::string &spelling_hint) const;
    void call_incoming_rc_strand(node_index node,
                                 const std::function<void(node_index, char)> &callback,
                                 const std::string &spelling_hint) const;

    // Find outgoing nodes that are on the reverse complement strand of node
    void adjacent_outgoing_rc_strand(node_index node,
                                     const std::function<void(node_index, uint64_t)> &callback,
                                     const std::string &spelling_hint) const;
    void call_outgoing_rc_strand(node_index node,
                                 const std::function<void(node_index, char)> &callback,
                                 const std::string &spelling_hint) const;
```

Reply:
So now `*_incoming_rc_strand` is called by `call_incoming_kmers` and `*_outgoing_rc_strand` is called by `call_outgoing_kmers`.